File: vm/home/tango-cs/examples/training/client-api/reading02.cpp
/*
 * reading02.cpp
 * 
 *  example program for Tango training
 * 
 * C. Scafuri
 * Elettra - Sincrotrone Trieste S.C.p.A.
 * 2016
 */



#include <tango.h>
#include <cstdlib>
using namespace Tango;
int main(int argc, char **argv)
{
    string devname = "sys/tg_test/1";
    string attname = "double_scalar";
    DeviceProxy *dev = 0;
    // step 1) create the device proxy
    try{
        dev = new Tango::DeviceProxy(devname);
    }
    catch(Tango::DevFailed &e)
    {
        Tango::Except::print_exception(e);
        cerr << "failed to create DeviceProxy" << endl;
        exit(EXIT_FAILURE); // semplicistic error handling
    }

    // step 2) prepare the reading
    DeviceAttribute da;
    double value=0.0;
    try{
        da = dev->read_attribute(attname); //read the attribute
        cout << "attibute_name:" << da.get_name() << endl;
        cout << "attibute_timestamp:" << da.get_date().tv_sec << endl;
        cout << "dim_x: "<< da.get_dim_x() << "  dim_y: "<<da.get_dim_y() << endl;
        cout << "type: "<< da.get_type() << endl;
        cout << "quality:" << da.get_quality() << endl;

        // get reading and setting , they (almost) always differ!!!
        if (da.get_dim_x() == 1 && da.get_dim_y() == 0){ //safety check on dimensions 
            vector<double> outval;
            da >> outval;  // the operator is now overloaed on vector<double>   != double !!!!!!
            cout << "reading:" << outval[0] << "  setting:" << outval[1] << endl;
        }
    }
    catch(DevFailed &e)
    {
        Tango::Except::print_exception(e);
        cerr << "failed to read attribute" << endl;
        delete dev; // free resources
        exit(EXIT_FAILURE); // semplicistic error handling
    }
    delete dev;
    exit(EXIT_SUCCESS);
}

File: vm/home/tango-cs/examples/training/python-api/reading02.py
#/!/usr/bin/python
#
# client example: reading an attribute , also with simplified syntax
# C. Scafuri - jan. 2016
#

from PyTango import *
import sys

devname = 'sys/tg_test/1'
attname = 'double_scalar'

#step 1) create the device proxy
try:
    dev = DeviceProxy(devname)
    print 'proxy for %s created' %(devname)
except DevFailed, ex:
        Except.print_exception(ex)
        sys.exit(1) #semplicistic error handling
        
#step 2 read the attribute
try:
    da = dev.read_attribute(attname) #read the attribute
    print 'attribute name:' , da.name
    print 'attibute_timestamp:', da.get_date()  # att.date
    print 'attribute_quality:', da.quality
    print 'attribute_dim_x:', da.dim_x
    print 'attribute_type:', da.type
    print 'attribute.value :' ,da.value
except DevFailed, ex:
    Except.print_exception(ex)
    print 'failed to read attribute'
    sys.exit(1) #simplicistic error handling

    #set a value with jive and check the result here
try:
    da = dev.read_attribute(attname) #read again the attribute
    print 'attribute.value    (reading):' ,da.value
    print 'attribute.w_value  (setting):' ,da.w_value  #simpler than C++
except DevFailed, ex:
    Except.print_exception(ex)
    print 'failed to read attribute'
    sys.exit(1) #simplicistic error handling
sys.exit(0)